### PR TITLE
Update print dialog to reuse existing dialog styles.

### PR DESCRIPTION
**Pull request by raymondwang**
Addresses [#3783](https://github.com/mozilla/pdf.js/issues/3783).

[Screenshot: http://cl.ly/2m3K3Z041M0o]

My understanding is that the print dialog was developed to be as modular as possible, with styles packed into the HTML, but this seemed counterintuitive to the issue and to reusability.

This commit is minor and I'm happy to do more if needed, including updating styles as desired. This is my first contribution to this project, so I didn't want to rock the boat.

### CODE REVIEW

File: web/viewer.css [hunk 1]
```diff
@@ -1428,10 +1429,33 @@ html[dir='rtl'] .outlineItemToggler::before {
...
+}
+
+#mozPrintCallback-shim progress {
+  width: 100%;
+}
+
+#mozPrintCallback-shim .relative-progress {
+  clear: both;
+  float: right;
+}
```
Comment (Snuffleupagus) [inline web/viewer.css]: Given that the `mozPrintCallback` is only included in `!(FIREFOX || MOZCENTRAL)` builds, I don't that we should place styles specific to it in `viewer.css` since those rules will be include for *all* builds. Hence I'd suggest that you move the above rules back.

File: web/viewer.css [hunk 2]
```diff
@@ -1419,6 +1419,7 @@ html[dir='rtl'] .outlineItemToggler::before {
 .dialog .buttonRow {
   text-align: center;
   vertical-align: middle;
+  clear: both;
```
Comment (Snuffleupagus) [inline web/viewer.css]: Why was this necessary?

File: web/viewer.html [hunk 3]
```diff
@@ -315,6 +315,9 @@
             </div>
           </div>
         </div>
+<!--#if !(FIREFOX || MOZCENTRAL)-->
+<!--#include viewer-snippet-mozPrintCallback-polyfill.html-->
+<!--#endif-->
```
Comment (Snuffleupagus) [inline web/viewer.html]: Is it actually necessary to move this block?

File: web/viewer-snippet-mozPrintCallback-polyfill.html [hunk 4]
```diff
@@ -1,79 +1,13 @@
-<div id="mozPrintCallback-shim" hidden>
-  <style>
-@media print {
-  #printContainer div {
```
Comment (Snuffleupagus) [inline web/viewer-snippet-mozPrintCallback-polyfill.html]: I'd expect that removing this line (as part of the code move that I suggested against), can actually cause issues.

File: web/viewer.css [hunk 5]
```diff
@@ -1428,10 +1429,33 @@ html[dir='rtl'] .outlineItemToggler::before {
...
+
 #passwordOverlay .toolbarField {
   width: 200px;
 }
 
+@media print {
+  #mozPrintCallback-shim {
+    display: none;
+    page-break-after: always;
+    page-break-inside: avoid;
```
Comment (Snuffleupagus) [inline web/viewer.css]: I'd expect that placing the `page-break-...` rules unconditionally in `viewer.css` will actually cause issues in e.g. FIrefox, since they were purposely moved in PR #5785.

File: web/viewer.css (hunk 1 above)
Comment (raymondwang) [inline web/viewer.css]: Fair enough, can do.

File: web/viewer.css (hunk 2 above)
Comment (raymondwang) [inline web/viewer.css]: Previously, `#mozPrintCallback-shim .progress-actions` required `clear: both;`. I wanted to be able to simply reuse this `.buttonRow` class, so it seemed straightforward enough for me to just lump it in here. It has no visible implications on anything else using this class, though if we're concerned about style property bloat I could make a new selector with more specificity.

File: web/viewer.css (hunk 5 above)
Comment (raymondwang) [inline web/viewer.css]: Sorry, was unaware of this. I'll throw this back in the snippet as well.

File: web/viewer.css (hunk 2 above)
Comment (Snuffleupagus) [inline web/viewer.css]: I suppose my question was more whether the `clear: both;` rule is actually necessary at all, now that you're using the new class.

**Comment by Rob--W** [comment]
First, thanks for wanting to contribute to PDF.js! 

> My understanding is that the print dialog was developed to be as modular as possible, with styles packed into the HTML, but this seemed counterintuitive to the issue and to reusability.

The current method makes reusability easier: other devs can copy the snippet and use it in their own projects (opposed to looking in viewer.css to figure out the styles).

File: web/viewer.css (hunk 2 above)
Comment (raymondwang) [inline web/viewer.css]: It did make a visual difference for the print dialog button, but I didn't spend any time investigating why that is. When I get home tonight I'll take another look!

**Comment by raymondwang** [comment]
I'm happy to help if I can! The documentation for this project is great; this is my first time trying to pitch in on something open source, and the README totally made the process less big and scary for me.

I see what you mean about reusability for the snippet; along with @Snuffleupagus's comment about keeping build-specific style rules in the snippet, I'll move these styles back into the HTML. As far as reusability of CSS styles, however, this seems like a little bit of a conflict. Would you prefer it if the snippet continued to live wholly on its own, with its own style rules? I can copy and paste the styles for the overlayContainer and dialog and write them into the scoped style tag, but these styles would definitely be less DRY as a result.

**Comment by Rob--W** [comment]
> As far as reusability of CSS styles, however, this seems like a little bit of a conflict. Would you prefer it if the snippet continued to live wholly on its own, with its own style rules? I can copy and paste the styles for the overlayContainer and dialog and write them into the scoped style tag, but these styles would definitely be less DRY as a result.

I suggest to keep the style at its current place. Moving it to a separate file doesn't add a clear benefit. Anyone who is interested in mozPrintCallback can easily tweak it as needed.

**Comment by raymondwang** [comment]
@Rob--W I've moved the styles back into the snippet, which can continue to live independently. I've basically just migrated each CSS rule that was being applied to the `#overlayContainer` and its relevant descendants, so it should stylistically resemble the password dialog now.

**Comment by Rob--W** [comment]
It seems that you've copied some styles from viewer.css and put it inline in the HTML. At the same time, the (font) size of the dialog decreased. Can you explain why this would be an improvement?

current

proposed

**Comment by raymondwang** [comment]
@Rob--W: I think I'm a little confused about the intent behind this ticket. Initially I thought that moving towards reusing classes from viewer.css would be the way to go, since the styles would be consistent with the other dialogs. However, you noted that you wanted this snippet to be able to live independently of viewer.css, which makes sense for reusability of this particular component, but not necessarily towards stylistic cohesion across dialogs. If I begin referencing viewer.css classes in the snippet, doesn't that defeat the purpose of keeping its styles inline and easily parseable as an independent component?

The compromise I went with here is just to explicitly copy over the styles from viewer.css inline (the font-size of the password dialog is 12px, which I personally think could be bumped up to 14px for both). I agree that this may not be necessary, but I'm not sure how else to fulfill both desires of making this dialog stylistically similar to the others, while also maintaining its independence as a reusable component. Again, I'm hopeful to contribute as much as makes sense, but I'm not exactly clear on what our goals are here.

**Comment by Rob--W** [comment]
@raymondwang It's fine if you only account for the PDF.js use case. Anyone who is going to use mozPrintCallback probably needs to customize the styles, so there is not a huge difference between deleting some CSS and modifying some class names.

In my opinion, the dialog's current style is OK (a decent look with minimal code). The only odd item is the button, whose style is not consistent with the rest of the viewer. And maybe the spacing between the progress bar and the progress number (percents) can be improved.

I'm not a visual designer, so if you have a better opinion then that's welcome.

**Comment by raymondwang** [comment]
Thanks for the reply (and for spending time walking me through this first contribution); I think I see where you're coming from now. I've opted to reuse dialog classes from viewer.css but to override any styles as necessary inline within the snippet. Notably, I've bumped the font-size to 14px as it was previously, and I also moved the padding from .progress-row to the progress bar, because that gives us some space which I agree was needed between the bar and .relative-progress. Let me know if this aligns more closely with what you're looking for. I think it makes more sense to me in terms of a step towards a more modular, reusable approach.

File: web/viewer-snippet-mozPrintCallback-polyfill.html [hunk 6]
```diff
@@ -1,79 +1,45 @@
...
-  }
-}
+    @media print {
+      #printContainer div {
+        page-break-after: always;
+        page-break-inside: avoid;
+      }
+    }
   </style>
   <style scoped>
```
Comment (Rob--W) [inline web/viewer-snippet-mozPrintCallback-polyfill.html]: `<style scoped>` only works in Firefox. Please insert the explicit `#mozPrintCallback-shim` selector back before each selector.

**Comment by timvandermeij** [comment]
I like how this PR is developing. The screenshot above shows exactly what I have in mind for the printing dialog, which is to make it look like the other dialogs such as the password prompt dialog. It is good to see that this also saves quite a bit of CSS in the snippet, and only overrides remain.

I'll leave further review to @Rob--W, but in general the patch looks good to me. Thank you for working on this, and welcome to the project!

**Comment by Rob--W** [comment]
The print dialog is now invisible. At the time of printing, the container (`#overlayContainer`) has `class="hidden"`, which prevents the child nodes (e.g. the print dialog) from being shown.

To test the visuals, I suggest to put a breakpoint at the `renderProgress` function in viewer.js.
In Chrome, this freezes the JS execution and adds a dark overlay. As a quick hack, you can also put a breakpoint at `canvas.mozPrintCallback` and run `canvas.mozPrintCallback=function(){}` to stop the print job from continuing (then the progress bar will be stuck at 6% and you can easily tweak the styles).

**Comment by timvandermeij** [comment]
@raymondwang Do you have enough information to improve this patch or do you need some assistance to complete it?

**Comment by raymondwang** [comment]
@timvandermeij @Rob--W Sorry, work has kept me busy the last month. I'm hoping to pick back up on this ticket soon. From what I understand about the codebase so far, being able to access the overlayContainer in order to change styles will probably require a rewriting of the shim to become a class, rather than static methods. I think it seems straightforward enough, but I'll have to scramble together some time in the near future to take care of it.

**Comment by timvandermeij** [comment]
No worries. Take your time and feel free to ask if you have any questions!

**Comment by timvandermeij** [comment]
Closing as superseded by #7697.
